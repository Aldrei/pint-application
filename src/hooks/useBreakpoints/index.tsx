import { useTheme } from '@mui/material/styles';
import useMediaQuery from '@mui/material/useMediaQuery';

export const useBreakpoints = () => {
  const theme = useTheme();

  const goSm = useMediaQuery(theme.breakpoints.up('sm'));
  const goMd = useMediaQuery(theme.breakpoints.up('md'));
  const goLg = useMediaQuery(theme.breakpoints.up('lg'));
  const goXl = useMediaQuery(theme.breakpoints.up('xl'));

  return [goSm, goMd, goLg, goXl];
};
