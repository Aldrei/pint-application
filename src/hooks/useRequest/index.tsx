import * as React from 'react';

import axios from '../../hooks/useConfigAxios';

// import { propertiesVideosService } from '../../services/properties';

import { IServiceRequestStatus } from '../../types';

interface IProps {
  type: 'propertyVideoDelete';
  code: string;
  videoId: string;
}

const useRequest = ({ code, videoId }: IProps) => {
  const [status, setStatus] = React.useState<IServiceRequestStatus>('idle');

  axios.get('https://swapi.dev/api/people', {
    onDownloadProgress: progressEvent => {
      setStatus('loading');
    }
  })
    .then((res) => {
      console.log('DEBUG RESPONSE:', res);
      setStatus('success');
    })
    .catch((err) => {
      console.log('DEBUG ERROR:', err);
      setStatus('failed');
    });

  return { status };
};

export default useRequest;