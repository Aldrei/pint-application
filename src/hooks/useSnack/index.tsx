import React from 'react';

import Snackbar from '@mui/material/Snackbar';
import MuiAlert, { AlertProps } from '@mui/material/Alert';

const Alert = React.forwardRef<HTMLDivElement, AlertProps>(function Alert(props, ref) {
  return <MuiAlert elevation={6} ref={ref} variant="filled" {...props} />;
});

const useSnack = () => {
  const [snacks, setSnacks] = React.useState(false);

  const handleClose = () => setSnacks(false);

  const SnackList = React.useCallback(() => {
    return (
      <Snackbar 
        open={snacks} 
        autoHideDuration={6000} 
        onClose={handleClose}
        sx={{ '& .MuiPaper-root': { flexDirection: 'row' } }}
      >
        <Alert 
          onClose={handleClose} 
          severity="success" 
          sx={{ width: '100%' }}
        >
          This is a success message!
        </Alert>
      </Snackbar>
    );
  }, [snacks]);

  const addSnack = React.useMemo(() => ({
    addMessage: () => {
      console.log('ADD SNACK MESSAGE!');
      setSnacks(!snacks);
    },
  }), []);

  return { addSnack, snacks, SnackList };
};

export default useSnack;