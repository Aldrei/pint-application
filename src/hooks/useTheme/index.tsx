import React from 'react';

import { createTheme as createThemeOrigin } from '@mui/material/styles';
import { PaletteMode } from '@mui/material';

declare module '@mui/material/styles' {
  interface Theme {
    pallete?: {
      mode: string;
      primary?: {
        main: string;
        dark: string;
      },
      background: {
        default: string;
        paper: string;
      },
      text: {
        primary: string;
        secondary: string;
      },
      divider: string;
    }
  }
  // allow configuration using `createTheme`
  interface ThemeOptions {
    pallete?: {
      mode: string;
      primary?: {
        main: string;
        dark: string;
      },
      background: {
        default: string;
        paper: string;
      },
      text: {
        primary: string;
        secondary: string;
      },
      divider: string;
    }
  }
}

const breakpoints = {
  values: {
    xs: 0,
    sm: 600,
    md: 900,
    lg: 1200,
    xl: 1536,
    mobile: 0,
    tablet: 640,
    laptop: 1024,
    desktop: 1200,
  },
};

const getDesignTokens = (mode: PaletteMode) => ({
  ...breakpoints,
  palette: {
    mode,
    ...(mode === 'light'
      ? {
        text: {
          primary: '#000000',
          secondary: 'rgb(172, 39, 189)'
        },
        background: {
          default: '#de1818',
          paper: '#1f10c6'
        },
        divider: '#c4e715'
      }
      : {
        text: {
          primary: 'rgb(255, 255, 255)',
          secondary: 'rgb(178, 186, 194)'
        },
        background: {
          default: 'rgb(0, 30, 60)',
          paper: 'rgb(10, 25, 41)'
        },
        divider: 'rgba(255, 255, 255, 0.12)'
      }
    ),
  },
});

export const createTheme = (mode: PaletteMode) => createThemeOrigin(getDesignTokens(mode));

const useTheme = () => {
  const [mode, setMode] = React.useState<'light' | 'dark'>('dark');

  const colorMode = React.useMemo(
    () => ({
      toggleColorMode: () => {
        setMode((prevMode) => (prevMode === 'light' ? 'dark' : 'light'));
      },
    }),
    [],
  );

  const theme = React.useMemo(() => createTheme(mode), [mode]);

  return { theme, colorMode };
};

export default useTheme;