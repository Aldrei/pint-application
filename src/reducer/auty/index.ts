import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { authService, IAuthServiceAccessTokenRequest, IAuthServiceAccessTokenResponse } from '../../services/auth';

import { IServiceRequest } from '../../types';

export interface IAutyState extends IServiceRequest {
  whoIsAuth: object;
  accessToken: IAuthServiceAccessTokenResponse;
}

const initialState: IAutyState = {
  status: 'idle',
  whoIsAuth: {},
  accessToken: {} as IAuthServiceAccessTokenResponse,
};

// The function below is called a thunk and allows us to perform async logic. It
// can be dispatched like a regular action: `dispatch(incrementAsync(10))`. This
// will call the thunk with the `dispatch` function as the first argument. Async
// code can then be executed and other actions can be dispatched. Thunks are
// typically used to make async requests.
export const authServiceThunk = createAsyncThunk(
  'counter/fetchCount',
  async (data: IAuthServiceAccessTokenRequest) => {
    const response = await authService.accessToken(data);
    // The value we return becomes the `fulfilled` action payload
    return response;
  }
);

export const authSlice = createSlice({
  name: 'auth',
  initialState,
  // The `reducers` field lets us define reducers and generate associated actions
  reducers: {},
  // The `extraReducers` field lets the slice handle actions defined elsewhere,
  // including actions generated by createAsyncThunk or in other slices.
  extraReducers: (builder) => {
    builder
      .addCase(authServiceThunk.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(authServiceThunk.fulfilled, (state, action) => {
        console.log('authServiceThunk fulfilled:', action);
        state.status = 'success';
        state.accessToken = action.payload?.data?.access_token ? action.payload.data : {};
        state.data = action.payload.data;
      })
      .addCase(authServiceThunk.rejected, (state, action) => {
        console.log('authServiceThunk rejected:', action);
        state.status = 'failed';
        state.data = action.payload as object;
        state.accessToken = {};
      });
  },
});

// The function below is called a selector and allows us to select a value from
// the state. Selectors can also be defined inline where they're used instead of
// in the slice file. For example: `useSelector((state: RootState) => state.counter.value)`
// export const selectCount = (state: RootState) => state.counter.value;

export default authSlice.reducer;
