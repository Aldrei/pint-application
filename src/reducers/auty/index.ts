import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { authService, IAuthServiceAccessTokenRequest, IAuthServiceAccessTokenResponse, IAuthServiceRevokeRequest } from '../../services/auth';
import { AppThunk } from '../../stores';

import { IServiceRequest, IWhoIsAuth } from '../../types';

export interface IAutyState extends IServiceRequest {
  whoIsAuth: IWhoIsAuth;
  accessToken: IAuthServiceAccessTokenResponse;
}

const initialState: IAutyState = {
  name: 'authReducer',
  status: 'idle',
  whoIsAuth: {} as IWhoIsAuth,
  accessToken: {} as IAuthServiceAccessTokenResponse,
};

export const authServiceThunk = createAsyncThunk(
  'auth/access_token',
  async (data: IAuthServiceAccessTokenRequest) => {
    const response = await authService.accessToken(data);
    // The value we return becomes the `fulfilled` action payload

    return response;
  }
);

export const revokeServiceThunk = createAsyncThunk(
  'auth/revoke',
  async (data: IAuthServiceRevokeRequest) => {
    const response = await authService.revoke(data);
    // The value we return becomes the `fulfilled` action payload

    return response;
  }
);

export const whoServiceThunk = createAsyncThunk(
  'auth/who_is_auth',
  async () => {
    const response = await authService.who();
    // The value we return becomes the `fulfilled` action payload

    return response;
  }
);

export const authSlice = createSlice({
  name: 'auth',
  initialState,
  // The `reducers` field lets us define reducers and generate associated actions
  reducers: {
    clearAccessToken: (state) => {
      state.accessToken = {} as IAuthServiceAccessTokenResponse;
      state.whoIsAuth = {} as IWhoIsAuth;
    }
  },
  // The `extraReducers` field lets the slice handle actions defined elsewhere,
  // including actions generated by createAsyncThunk or in other slices.
  extraReducers: (builder) => {
    builder
      /** Access token */
      .addCase(authServiceThunk.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(authServiceThunk.fulfilled, (state, action) => {
        state.status = 'success';
        state.accessToken = action.payload?.data?.access_token ? action.payload.data : {};
        state.data = action.payload.data;
      })
      .addCase(authServiceThunk.rejected, (state, action) => {
        state.status = 'failed';
        state.data = action.payload as object;
        state.accessToken = {};
      })
      /** Revoke. */
      .addCase(revokeServiceThunk.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(revokeServiceThunk.fulfilled, (state, action) => {
        state.status = 'success';
        state.data = action.payload.data;
        state.accessToken = {};
        state.whoIsAuth = {} as IWhoIsAuth;
      })
      .addCase(revokeServiceThunk.rejected, (state, action) => {
        state.status = 'failed';
        state.data = action.payload as object;
      })
      /** Who is auth. */
      .addCase(whoServiceThunk.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(whoServiceThunk.fulfilled, (state, action) => {
        state.status = 'success';
        state.whoIsAuth = action.payload.data;
      })
      .addCase(whoServiceThunk.rejected, (state, action) => {
        state.status = 'failed';
        state.whoIsAuth = action.payload as IWhoIsAuth;
      });
  },
});

const { clearAccessToken } = authSlice.actions;

export const clearAccessTokenReducer = (): AppThunk => (dispatch) => {
  dispatch(clearAccessToken());
};

export default authSlice.reducer;
