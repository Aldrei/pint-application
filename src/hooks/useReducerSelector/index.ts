import { RootState, IReducersType } from '../../stores';
import { useAppSelector, useSelectorCallback } from '../../stores/hooks';

export const useAppSelectorBlaBlaBal = (reducer: keyof IReducersType) => useAppSelector((state: RootState) => useSelectorCallback(state, reducer));