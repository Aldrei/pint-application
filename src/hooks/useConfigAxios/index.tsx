import axios from 'axios';

import { clearAccessTokenReducer } from '../../reducers/auty';
import { useAppDispatch } from '../../stores/hooks';

const api = axios.create({
  baseURL: process.env.REACT_APP_API_BASE_URL,
});

const getToken = () => {
  try {
    const persistImob = localStorage.getItem('persist:imob');
    if (persistImob) {
      const persistImobData = JSON.parse(persistImob);

      if (persistImobData.authReducer) {
        const authReducerData = Object.assign({}, JSON.parse(persistImobData.authReducer));
        return authReducerData?.accessToken?.access_token;
      }
    }

    return '';
  } catch (error) {
    console.log(error);
    return '';
  }
};


// Add a request interceptor
api.interceptors.request.use(function (config) {
  if (config.headers) config.headers['Authorization'] = `Bearer ${getToken()}`;
  return config;
}, function (error) {
  // Do something with request error
  return Promise.reject(error);
});

export const useConfigAxios = () => {
  const dispatch = useAppDispatch();

  // Add a response interceptor
  api.interceptors.response.use((response) => {
    console.log('api.interceptors.response success...');
    // Any status code that lie within the range of 2xx cause this function to trigger
    /** NOTE: Suppress 'A non-serializable value was detected in the state, in the path' error */
    return JSON.parse(JSON.stringify(response));
  }, (error) => {
    console.log('api.interceptors.response error...', { error });
    // Any status codes that falls outside the range of 2xx cause this function to trigger
    // return Promise.reject(error);

    if (error.response.status === 401) {
      dispatch(clearAccessTokenReducer());
    }

    /** NOTE: Suppress 'A non-serializable value was detected in the state, in the path' error */
    return JSON.parse(JSON.stringify(error.response));
  });

  return [];
};

export default api;