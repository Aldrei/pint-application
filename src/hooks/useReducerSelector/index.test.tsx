import { useAppSelectorBlaBlaBal } from './index';

import { useSelector } from 'react-redux';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

import { IAutyState } from '../../reducers/auty';
import { ICounterState } from '../../reducers/counter';
import { useAppSelector } from '../../stores/hooks';

jest.mock('../../store/hooks', () => ({
  useAppSelector: jest.fn(),
}));

describe('Helper dataFormControl', () => {
  const reducerValidFlag = 'authReducer';

  const useSelectorMocked = useSelector as jest.MockedFunction<typeof useSelector>;
  const useAppSelectorMocked = useAppSelector as jest.MockedFunction<typeof useAppSelector>;

  it('Should return a valid flag reducer', () => {
    const authReducerValueRoot: IAutyState = {
      whoIsAuth: {}, 
      accessToken: {},
      status: 'idle'
    };
    const counterReducerValueRoot: ICounterState = {
      value: 123, 
      status: 'idle',
    };
    const rootReducerValue = { authReducer: authReducerValueRoot, counterReducer: counterReducerValueRoot };
    useSelectorMocked.mockReturnValue(rootReducerValue);

    const authReducerValue: IAutyState = {
      whoIsAuth: {}, 
      accessToken: {},
      status: 'idle'
    };
    useAppSelectorMocked.mockReturnValue(authReducerValue);

    const objValues = useAppSelectorBlaBlaBal(reducerValidFlag);

    expect(objValues).toEqual(authReducerValueRoot);
  });
});
