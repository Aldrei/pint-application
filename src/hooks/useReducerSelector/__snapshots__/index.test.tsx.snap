// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helper dataFormControl Should return authReducer data 1`] = `
<body>
  <div>
    <span>
      ReducerName: authReducer
    </span>
  </div>
</body>
`;

exports[`Helper dataFormControl Should return propertiesListReducer data 1`] = `
<body>
  <div>
    <span>
      ReducerName: propertiesListReducer
    </span>
  </div>
</body>
`;

exports[`Helper dataFormControl Should return propertiesPhotosReducer data 1`] = `
<body>
  <div>
    <span>
      ReducerName: propertiesPhotosReducer
    </span>
  </div>
</body>
`;

exports[`Helper dataFormControl Should return propertiesShowReducer data 1`] = `
<body>
  <div>
    <span>
      ReducerName: propertiesShowReducer
    </span>
  </div>
</body>
`;
